### PR TITLE
Fixes for shader support

### PR DESCRIPTION
**Pull request by krasnotsvetov**

### CODE REVIEW

File: resharper/resharper-unity/src/HlslSupport/ShaderLabInclusionContextProvider.cs [hunk 1]
```diff
@@ -30,11 +31,14 @@ public class ShaderLabInclusionContextProvider : IInclusionContextProvider
...
             };
 
+            properties.PredefinedMacros.Add(CppPPDefineSymbol.ParsePredefinedMacro("SHADER_API_D3D11"));
             var cgIncludeFolder =
                 CgIncludeDirectoryTracker.GetCgIncludeFolderPath(cache.Solution.GetComponent<UnityVersion>());
             if (!cgIncludeFolder.IsEmpty)
             {
                 properties.IncludePaths.Add(cgIncludeFolder);
+                properties.IncludePaths.Add(cache.Solution.SolutionDirectory);
+                properties.IncludePaths.Add(cache.Solution.SolutionDirectory.Combine("Library"));
```
Comment (citizenmatt) [inline resharper/resharper-unity/src/HlslSupport/ShaderLabInclusionContextProvider.cs:41]: Will adding solution directory handle includes starting with `Assets/`?

Is there anything under `Library` that can be included, or is that just for `Library/PackageCache`?

File: resharper/resharper-unity/src/Application/UI/Options/VisualStudio/ReSharperOptionsPage.cs [hunk 2]
```diff
@@ -64,6 +64,11 @@ public class ReSharperOptionsPage : BeSimpleOptionsPage
             AddBoolOption((UnitySettings s) => s.IsAssetIndexingEnabled,
                 "Parse text based asset files for implicit script usages (requires re-opening solution)");
 
+            
+            AddHeader("Shaders");
+            AddBoolOption((UnitySettings s) => s.SuppressShaderErrorHighlighting,
+                "Suppress resolve errors of unqualified names in HLSL");
```
Comment (citizenmatt) [inline resharper/resharper-unity/src/Application/UI/Options/VisualStudio/ReSharperOptionsPage.cs:70]: Why "HLSL" here, when we're currently only checking `CGPROGRAM` and `CGINCLUDE`?

File: resharper/resharper-unity/src/HlslSupport/UnityGameEngineDirectiveResolver.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            mySolution = solution;
+            mySolutionTracker = solutionTracker;
+            myReferencesTracker = referencesTracker;
+            myLogger = logger;
+        }
+        
+        public bool IsApplicable(CppInclusionContext context, string path)
+        {
+              return path.StartsWith("Packages/") &&
+                   (mySolutionTracker.IsUnityProject.HasTrueValue() || myReferencesTracker.HasUnityReference.HasTrueValue());
```
Comment (citizenmatt) [inline resharper/resharper-unity/src/HlslSupport/UnityGameEngineDirectiveResolver.cs:37]: Do we need to do anything to support includes relative to `Assets/` here?

File: resharper/resharper-unity/src/HlslSupport/UnityGameEngineDirectiveResolver.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                                {
+                                    if (!folder.IsDirectory)
+                                        continue;
+                                        
+                                    var packageAndVersion = folder.RelativePath.Name.Split('@');
+                                    if (packageAndVersion.Length != 2)
+                                        continue;
+                                    
+                                    if (result.ContainsKey(packageAndVersion[0]))
+                                        continue;
```
Comment (citizenmatt) [inline resharper/resharper-unity/src/HlslSupport/UnityGameEngineDirectiveResolver.cs:93]: You could use `JetSemanticVersion` on the second element and always use the highest version.

File: resharper/resharper-unity/src/HlslSupport/UnityHlslCppCompilationPropertiesProvider.cs [hunk 5]
```diff
@@ -28,6 +29,10 @@ public UnityHlslCppCompilationPropertiesProvider(UnityVersion unityVersion)
                 if (!path.IsEmpty)
                     properties.IncludePaths.Add(path);
 
+                properties.PredefinedMacros.Add(CppPPDefineSymbol.ParsePredefinedMacro("SHADER_API_D3D11"));
+                properties.IncludePaths.Add(globalCache.Solution.SolutionDirectory);
+                properties.IncludePaths.Add(globalCache.Solution.SolutionDirectory.Combine("Library"));
```
Comment (citizenmatt) [inline resharper/resharper-unity/src/HlslSupport/UnityHlslCppCompilationPropertiesProvider.cs:34]: Same comment as `ShaderLabInclusionContextProvider` - do we have shaders that are relative to `Library`?

File: resharper/resharper-unity/src/HlslSupport/ShaderLabInclusionContextProvider.cs (hunk 1 above)
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/HlslSupport/ShaderLabInclusionContextProvider.cs:41]: It is just for package cache, we could remove PackageCache and add to search path Library/PackageCache and Packages folder, but then we will need to add unity specific logic for determing priority of that folders to cpp logic, thus I decide to add Library as seperated include source folder

File: resharper/resharper-unity/src/Application/UI/Options/VisualStudio/ReSharperOptionsPage.cs (hunk 2 above)
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/Application/UI/Options/VisualStudio/ReSharperOptionsPage.cs:70]: Not only CGPROGRAM and CGINCLUDE, we also support HLSLPROGRAM/HLSLINCLUDE, .hlsl files, .cg files. All that files use HLSL language

File: resharper/resharper-unity/src/HlslSupport/UnityHlslCppCompilationPropertiesProvider.cs (hunk 5 above)
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/HlslSupport/UnityHlslCppCompilationPropertiesProvider.cs:34]: Same answer :-)

File: resharper/resharper-unity/src/HlslSupport/UnityGameEngineDirectiveResolver.cs (hunk 3 above)
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/HlslSupport/UnityGameEngineDirectiveResolver.cs:37]: No, when shader is used from Assets folder, it uses real file system path related to solution folder (https://youtrack.jetbrains.com/issue/RIDER-47412)

**Pull request merged**